tests/unit_tests/c_wrappers/res/config/test_config_parser.py

```diff
@@ -313,3 +314,47 @@ def test_valid_string():
     assert ContentTypeEnum.CONFIG_BOOL.convert_string("True")
     assert not ContentTypeEnum.CONFIG_BOOL.convert_string("False")
     assert not ContentTypeEnum.CONFIG_BOOL.convert_string("F")
+
+
+@pytest.mark.parametrize(
+    "option_list, offset, expected",
+    [
+        (["a:1"], 0, {"a": "1"}),
+        (["B:abc"], 0, {"B": "abc"}),
+        (["A"], 0, {}),
+        (["a:1", "B:abc"], 0, {"a": "1", "B": "abc"}),
+        (["a:1", "T:"], 0, {"a": "1"}),
+        (["a:1", ":T"], 0, {"a": "1"}),
+        (["a:1", ":-:"], 0, {"a": "1"}),
+        (["a:1", "T:"], 0, {"a": "1"}),
+        (["a:1", "b:2", "c:3"], 2, {"c": "3"}),
+        (["a", "b", "c"], 0, {}),
+    ],
+)
+def test_options_dic(option_list, offset, expected):
+    assert _option_dict(option_list, offset) == expected
```

Comment:
> i've read at some point that one shouldn't necessarily test implementation details, or every method / function that one writes. it was argued that this kind of test might be nice to have during development (in particular employing TDD), to make sure that the function turns out alright, but can be removed afterwards.

I personally did not read this at any point, and are you arguing now that simple functionality unit tests can/should be removed? Or am I missing the point? 

> when / if we ever change the implementation of ensemble config, or how the options are parsed, this test will have to be rewritten / updated as well, and that's an unnecessary slow down, some would say.

Those that say "that's an unnecessary slow down" are focusing too much on writing code just for the sake of writing a lot of code, not really focusing on catching errors before they hit the users.

>one might be able to make sure that this parsing works correctly by supplying enough cases via a config file / dict, using the hypothesis approach - and perhaps write a few non-correct examples (again on the level of config dict / file), to make sure that it errors / ignores in the expected way.

I agree that the test can be improved and should also account for more edge case like when the list provided is empty or if the list has non-string elements but in those cases, I assumed that would result in an invalid config file, and the parsing of the config file should handle it. But I will make a not to improve the test.

Reply:
ok i remember now that i heard it, it was a talk eivind linked to on slack - https://equinor.slack.com/archives/C3HLP8ZTQ/p1659420684498539 [internal link], the talk is https://youtu.be/EZ05e7EMOLM - there's a lot of introduction / background in the beginning, which might still be worth a listen, but the main message i'm referring to start at around timestamp 24:00

> are you arguing now that simple functionality unit tests can/should be removed?

yeah i know it sounds crazy. but my point (also with the "higher level approach") is this. if we have tests that pass in config files or dict with contents that cover the cases you have in the unit tests (by providing the necessary lines / key-values that will hit exactly those functions the unit tests are testing with the right arguments), then we get the same coverage - but we don't need to worry about how we implemented it, and are free to change details about those small methods later, or refactor the way its implemented (as in for instance use some other methods to achieve the same thing), without having to change our tests.

i hope this makes it a bit clearer, if not, we can disregard it here, and i will try to bring it up in a retrospective instead, or some other forum.

Reply:
btw i was also kinda shocked when i was presented with the notion to stop writing this kind of unit tests, as that was how i was literally introduced to TDD... and it seems so basic and logical, to get the safety of (trying to) covering each method you have, then you gotta be good, right? so it seems nonsensical when somebody says never mind, we don't need those...

Reply:
Thing is that one persons implementation detail is someone elses feature. Just because this function is private to ensemble_config, it is still valid to test it in that domain, but it would be good to test it through a public interface. However, I feel this is a good time to instead use doctest as the problem here is that these tests are weird when out of context. I would much rather have these as part of the docstring, and then testsed through `pytest --doctest-modules`.

Ie 

```
def _str_to_bool(text):
  """
  This function converts truthy text to boolean values according to the 
  rules of the FORWARD_INIT keyword, we keep these rules for backwards compatability

  First, any upper/lower case true/false value is converted to the corresponding boolean value:
  
  >>> _str_to_bool("TRUE")
  True
  >>> _str_to_bool("true")
  True
  
  ...
```

Reply:
Separately: is logging the failure the correct behavior here? We have this `ConfigValidationError` we raise to tell the user that there was an error in the configuration file now.

Reply:
ohhh doctests time to learn something new :heart_eyes:

Reply:
Work on adding doctest is done here: https://github.com/equinor/ert/pull/4278